### ISSUE TITLE
Assignment to final variables is allowed in a match statement

### ISSUE THREAD

**Issue by omerhadari** [python/mypy#19507]
**Bug Report**

When matching patterns assignment can override Final variables, and `mypy` does not report an error.

**To Reproduce**

```python
from typing import Final
FOO: Final = 8

a = 10

match a:
    case FOO:
        pass
print(FOO)  # FOO is reassigned, prints 10
```

**Expected Behavior**

For MyPy to report an error because `FOO` is final and this assignment should be disallowed.

**Actual Behavior**

No error is reported
**Your Environment**

Tested on MacOS and Ubuntu (wsl).

- Mypy version used: 1.17.0
- Mypy command-line flags: none
- Python version used: 3.11.3
